In [1]:
#imports
import pandas as pd

#file
filename = "Resources/purchase_data.csv"

#load file into dataframe
purchase_data = pd.read_csv(filename)
purchase_data.sample(1)

Purchase ID           SN  Age Gender  Item ID  \
290          290  Idastidru52   24   Male      147   

                             Item Name  Price  
290  Hellreaver, Heirloom of Inception   4.93

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [2]:
purchases_df = pd.DataFrame({
    "Number of Unique Customers":[len(purchase_data['SN'].unique())],
    "Number of Unique Items":[len(purchase_data['Item ID'].unique())],
    "Average Purchase Price":[round(purchase_data['Price'].mean(),2)],
    "Total Number of Purchases":[purchase_data['Purchase ID'].count()],
    "Total Revenue":[purchase_data['Price'].sum()]
})
purchases_df

Number of Unique Customers  Number of Unique Items  Average Purchase Price  \
0                         576                     183                    3.05   

   Total Number of Purchases  Total Revenue  
0                        780        2379.77

## Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

In [3]:
customers = purchase_data.drop_duplicates(subset='SN').groupby(['Gender'])['SN'].count()
perc_customers = round(customers/customers.sum()*100,2)
gender_demographics_df = pd.DataFrame({
    "Number of Players":customers,
    "% of Players":perc_customers
})
gender_demographics_df

Number of Players  % of Players
Gender                                                
Female                                81         14.06
Male                                 484         84.03
Other / Non-Disclosed                 11          1.91


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [4]:
#purchase count by gender
p_counts = purchase_data.groupby('Gender')['Purchase ID'].count()

#average purchase price by gender
p_gen = round(purchase_data.groupby('Gender')['Price'].mean(),2)

#toal purchase value by gender
gen_totals = purchase_data.groupby('Gender')['Price'].sum()

#average purchase total per person by gender
#i interpret this as the averages of all purchases by each man, woman, and other
gender_group = purchase_data.groupby(['Gender','SN'])['Price'].sum()
averages_gender=round(gender_group.groupby(['Gender']).mean(),2)

#average item price by gender
item_gender = round(purchase_data.groupby(['Gender'])['Price'].mean(),2)

gender_analysis_df = pd.DataFrame({
    "Purchase Count":p_counts,
    "Avg Purchase Price":p_gen,
    "Total Purchase Value":gen_totals,
    "Avg Purchase Total per Person":averages_gender,
    "Avg Item Price":item_gender
})
gender_analysis_df

Purchase Count  Avg Purchase Price  \
Gender                                                      
Female                            113                3.20   
Male                              652                3.02   
Other / Non-Disclosed              15                3.35   

                       Total Purchase Value  Avg Purchase Total per Person  \
Gender                                                                       
Female                               361.94                           4.47   
Male                                1967.64                           4.07   
Other / Non-Disclosed                 50.19                           4.56   

                       Avg Item Price  
Gender                                 
Female                           3.20  
Male                             3.02  
Other / Non-Disclosed            3.35

## Age Demographics

* Establish bins for ages
* Categorize the existing players using the age bins. Hint: use pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Optional: round the percentage column to two decimal points
* Display Age Demographics Table

In [5]:
bins = [0,9,14,19,24,29,34,39,99]
bin_names = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '>39']
purchase_data['Age Group'] = pd.cut(purchase_data['Age'], bins, labels=bin_names)
ages = purchase_data.drop_duplicates(subset='SN').groupby(['Age Group'])['SN'].count()
perc_ages = round(ages/ages.sum()*100,2)
age_demo = pd.DataFrame({
    "Number of Players":ages,
    "% of Players":perc_ages
})
age_demo

Number of Players  % of Players
Age Group                                 
<10                       17          2.95
10-14                     22          3.82
15-19                    107         18.58
20-24                    258         44.79
25-29                     77         13.37
30-34                     52          9.03
35-39                     31          5.38
>39                       12          2.08

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age
* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame

In [6]:
#purchase count
p_counts_age = purchase_data.groupby('Age Group')['Purchase ID'].count()

#average purchase price
p_age = round(purchase_data.groupby('Age Group')['Price'].mean(),2)

#total purchase value
age_totals = purchase_data.groupby('Age Group')['Price'].sum()

#average purchase total per person by age group
#i interpret this as the average of all purchases by each person in each age group
age_group = purchase_data.groupby(['Age Group','SN'])['Price'].sum().dropna()
averages_age = round(age_group.groupby('Age Group').mean(),2)

item_age = round(purchase_data.groupby('Age Group')['Price'].mean(),2)

age_purch = pd.DataFrame({
    "Purchase Count":p_counts_age,
    "Avg Purchase Price":p_age,
    "Total Purchase Value":age_totals,
    "Avg Purchase Total per Person":averages_age,
    "Avg Item Price":item_age
})
age_purch

Purchase Count  Avg Purchase Price  Total Purchase Value  \
Age Group                                                             
<10                    23                3.35                 77.13   
10-14                  28                2.96                 82.78   
15-19                 136                3.04                412.89   
20-24                 365                3.05               1114.06   
25-29                 101                2.90                293.00   
30-34                  73                2.93                214.00   
35-39                  41                3.60                147.67   
>39                    13                2.94                 38.24   

           Avg Purchase Total per Person  Avg Item Price  
Age Group                                                 
<10                                 4.54            3.35  
10-14                               3.76            2.96  
15-19                               3.86            3.04  
20-24                               4.32            3.05  
25-29                               3.81            2.90  
30-34                               4.12            2.93  
35-39                               4.76            3.60  
>39                                 3.19            2.94

## Top Spenders

* Run basic calculations to obtain the results in the table below
* Create a summary data frame to hold the results
* Sort the total purchase value column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [7]:
#identify top 5 spenders by total purchase value,
spenders = list(purchase_data.groupby('SN')['Price'].sum().sort_values().tail().index)

#list SN, purchase count, averae purchase price, and total purchase value
spender_purchases = purchase_data[purchase_data['SN'].isin(spenders)]
spender_avgs = round(spender_purchases.groupby(['SN'])['Price'].mean(),2)
spender_counts = spender_purchases.groupby(['SN'])['Price'].count()
spender_totals = spender_purchases.groupby(['SN'])['Price'].sum()

top_spenders = pd.DataFrame({
    "Purchase Count":spender_counts,
    "Avg Purchase Price":spender_avgs,
    "Total Purchase Value":spender_totals
})
top_spenders

Purchase Count  Avg Purchase Price  Total Purchase Value
SN                                                                   
Chamjask73                3                4.61                 13.83
Idastidru52               4                3.86                 15.45
Iral74                    4                3.40                 13.62
Iskadarya95               3                4.37                 13.10
Lisosia93                 5                3.79                 18.96

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns
* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
* Create a summary data frame to hold the results
* Sort the purchase count column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [8]:
#most popular items by purchase count
popular_items = list(purchase_data.groupby('Item ID')['Item ID'].count().sort_values().tail().index)
top_sellers = purchase_data[purchase_data['Item ID'].isin(popular_items)]
names = top_sellers.drop_duplicates(subset='Item Name').set_index('Item ID')['Item Name']
top_counts = top_sellers.groupby('Item ID')['Purchase ID'].count()
top_prices = top_sellers.groupby('Item ID')['Price'].max()
top_totals = top_sellers.groupby('Item ID')['Price'].sum()
top_items = pd.DataFrame({
    "Item Name":names,
    "Purchase Count":top_counts,
    "Item Price":top_prices,
    "Total Purchase Value":top_totals
})
top_items

Item Name  Purchase Count  \
Item ID                                                                 
82                                            Nirvana               9   
92                                       Final Critic               8   
108         Extraction, Quickblade Of Trembling Hands               9   
145                              Fiery Glass Crusader               9   
178      Oathbreaker, Last Hope of the Breaking Storm              12   

         Item Price  Total Purchase Value  
Item ID                                    
82             4.90                 44.10  
92             4.88                 39.04  
108            3.53                 31.77  
145            4.58                 41.22  
178            4.23                 50.76

## Most Profitable Items

* Sort the above table by total purchase value in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the data frame

In [9]:
#most profitable items by total purchase value
profit_items = list(purchase_data.groupby('Item ID')['Price'].sum().sort_values().tail().index)
profit_items
top_profits = purchase_data[purchase_data['Item ID'].isin(profit_items)]
profit_names = top_profits.drop_duplicates(subset='Item Name').set_index('Item ID')['Item Name']
profit_counts = top_profits.groupby('Item ID')['Purchase ID'].count()
profit_prices = top_profits.groupby('Item ID')['Price'].max()
profit_totals = top_profits.groupby('Item ID')['Price'].sum()
most_profitable_items = pd.DataFrame({
    "Item Name":profit_names,
    "Purchase Count":profit_counts,
    "Item Price":profit_prices,
    "Total Purchase Value":profit_totals
}).sort_values(by='Total Purchase Value',ascending=False)
most_profitable_items

Item Name  Purchase Count  \
Item ID                                                                 
178      Oathbreaker, Last Hope of the Breaking Storm              12   
82                                            Nirvana               9   
145                              Fiery Glass Crusader               9   
92                                       Final Critic               8   
103                                    Singed Scalpel               8   

         Item Price  Total Purchase Value  
Item ID                                    
178            4.23                 50.76  
82             4.90                 44.10  
145            4.58                 41.22  
92             4.88                 39.04  
103            4.35                 34.80

## Notes

* Item ID 103 is in the top 5 most profitable but not the top 5 purchase count. Put it on sale.
* The opposite is true for Item ID 108. Raise its price to increase profits.
* Send coupons for popular/profitable items to top spenders, but take note of their past purchases:

In [10]:
spender_purchases[spender_purchases['Item ID'].isin(set(profit_items)|set(popular_items))]

Purchase ID           SN  Age  Gender  Item ID  \
222          222   Chamjask73   22  Female      178   
247          247  Iskadarya95   20    Male       82   

                                        Item Name  Price Age Group  
222  Oathbreaker, Last Hope of the Breaking Storm   4.23     20-24  
247                                       Nirvana   4.90     20-24